src/libopensc/card-piv.c

```diff
@@ -3064,38 +3206,105 @@ static int piv_match_card_continued(sc_card_t *card)
 	}
 
 	/*
-	 * detect if active AID is PIV. NIST 800-73 says Only one PIV application per card
-	 * and PIV must be the default application
-	 * This can avoid doing doing a select_aid and losing the login state on some cards
+	 * Detect if active AID is PIV. NIST 800-73 says only one PIV application per card
+	 * and PIV must be the default application.
+	 * Try to avoid doing a select_aid and losing the login state on some cards.
 	 * We may get interference on some cards by other drivers trying SELECT_AID before
-	 * we get to see if PIV application is still active.
+	 * we get to see if PIV application is still active
 	 * putting PIV driver first might help. 
-	 * This may fail if the wrong AID is active
+	 * This may fail if the wrong AID is active.
+	 * Discovery Object introduced in 800-73-3 so will return 0 if found and PIV applet active.
+	 * Will fail with SC_ERROR_FILE_NOT_FOUND if 800-73-3 and no Discovery object.
+	 * But some other card could also return SC_ERROR_FILE_NOT_FOUND.
+	 * Will fail for other reasons if wrong applet is selected, or bad PIV implimentation. 
 	 */
-	i = piv_find_discovery(card);
+	
+	sc_debug(card->ctx,SC_LOG_DEBUG_MATCH, "PIV_MATCH card->type:%d CI:%08x r:%d\n", card->type,  priv->card_issues, r);
+	if (priv->card_issues & CI_DISCOVERY_USELESS) /* TODO may be in wrong place */
+		i = -1;
+	else
+		i = piv_find_discovery(card);
 
+	sc_debug(card->ctx,SC_LOG_DEBUG_MATCH, "PIV_MATCH card->type:%d i:%d CI:%08x r:%d\n", card->type, i, priv->card_issues, r);
 	if (i < 0) {
 		/* Detect by selecting applet */
 		i = piv_find_aid(card);
 	}
 
+	sc_debug(card->ctx,SC_LOG_DEBUG_MATCH, "PIV_MATCH card->type:%d i:%d CI:%08x r:%d\n", card->type, i, priv->card_issues, r);
 	if (i >= 0) {
+		 /* We now know PIV AID is active, test CCC object  800-73-* say CCC is required */
+		switch (card->type)  {
+			int iccc = 0;
```

Comment:
I can not build the source with the `gcc (GCC) 8.2.1 20181105 (Red Hat 8.2.1-5)`, which is causing the following error:
```
card-piv.c: In function 'piv_match_card_continued':
card-piv.c:3271:8: error: statement will never be executed [-Werror=switch-unreachable]
    int iccc = 0;
        ^~~~
```

Reply:
Should be fixed, moved the declare statement.